### ISSUE TITLE
Fill match prioritises unimported Type over imported Type when names are the same

### ISSUE THREAD

**Issue by jroddev** [rust-lang/rust-analyzer#12709]
Diagnostics correctly shows Direction as imported - Up, Down, Left, Right
This is Direction from crate::core::Direction
![Screenshot_20220707_192035](https://user-images.githubusercontent.com/9654410/177740871-8831c70d-2164-4b69-b7a9-e3485ee0f254.png)

Code Action incorrectly fills match with bevy::ui::Direction - Inherit, LeftToRight, RightToLeft
Bevy is included in this project, but not imported in this file.
![Screenshot_20220707_192124](https://user-images.githubusercontent.com/9654410/177740887-d8b4dd82-f320-4a26-a410-44236fea3eee.png)

- match works correctly if bevy is not included in the project (even if installed on system)
- match works correctly in the file that my Direction is declared

----

NeoVim with LSP
**rust-analyzer version**: rust-analyzer 0.0.0 (427061da1 2022-06-19)
**rustc version**: rustc 1.62.0 (a8314ef7d 2022-06-27)

**Comment by flodiebold** [comment]
Can you provide the code for this, or a separate reproduction?

**Comment by jroddev** [comment]
This is the repo it occurs in for me: https://github.com/jroddev/bevy_snake
I open src/food.rs in neovim with lsp+rust_analyzer
add this piece of code
```
use crate::core::Direction;

fn use_dir(dir: &Direction) {
    match dir {
    }
}
```
instead of filling with `create::core::Direction` it fills with `bevy::ui::Direction`.
CLion fills `crate::core::Direction` here as expected.


I have tried to reproduce the issue in a smaller repository but it works as expected there.

**Comment by Veykril** [comment]
Ah there is a `use bevy::prelude::*;` in this file, I imagine some part of the code is getting confused and picks the glob imported `Direction` over the explicit import...

**Comment by Veykril** [comment]
Looked into this a bit, the following is a repro:
```rs
//- minicore: derive
//- /lib.rs crate:lib deps:dep
use crate::z::Direction;

mod z {
    use dep::*;

    #[derive()]
    pub struct Direction {}
             //^^^^^^^^^
}

fn use_dir(dir: &Direction$0) {}

//- /dep.rs crate:dep
pub struct Direction {}
```
This goto_def test fails, the important part is the `derive`, remove that and it succeeds...

**Comment by Veykril** [comment]
Fixed by #13602

**Comment by lowr** [comment]
Oh my, how could I forget to link this when it is the very issue I started with :sweat_smile: 

@Veykril Thanks for catching it, and also for your reduced repro, it helped a ton!
